Cache entities contribute main

AbhinavS96 opened the pull request:
## Description

Fixes #12511

This PR will boost the performance of relationships by caching the database tables that store the entity types / relationship types. These tables are rarely updated & queried often. 

## Instructions for Reviewers
We tested this by running a custom script and compared the performance with and without the changes applied. Loop over a list of entities/relationshiptypes and use `entityTypeService.findByEntityType(context, entity)` or `relationshipTypeService.findByLeftwardOrRightwardTypeName(context, entity)`. Creating a new context intermittently will also force the use of cache better. Summing up the time the above snippet took and comparing will show the performance improvement. Sample snippet: 

```
finalEndTime[0] = 0;
for (int i = 0; i < 500; i++) {
    List.of("isAuthorOfPublication", "isPublicationOfAuthor").forEach((entity) -> {
        try {
            Context context = new Context(Context.Mode.BATCH_EDIT);
            context.turnOffAuthorisationSystem();
            long startTime = System.currentTimeMillis();
            relationshipTypeService.findByLeftwardOrRightwardTypeName(context, entity);
            long endTime = System.currentTimeMillis();
            finalEndTime[0] += (endTime - startTime);
            context.complete();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    });
}
```

List of changes in this PR:
* We made EntityType / RelationshipType database entities cacheable
* We cache the queries that query these entities
* We added an uppercase index on the entity_type table, since the only query on that label queries case intensive by using an uppercase

Where the performance hits will become noticeable:
* When (bulk) creating relationships
* When retrieving metadata of an item with relationships
* Implicitly 
    * when updating metadata of an item with relationships (because we get the metadata before we update)
    * When creating a new version of an item with a lot of relationships (because the creating of relationships received a performance boost)

## Checklist

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

Commit:
Improve caching for entities and relationship types

Commit:
add expiry to ehcache-config

Commit:
remove unsupported upper() from H2 migration

Commit:
fix cache concurrency strategy

Commit:
Merge remote-tracking branch 'atmire-upstream/w2p-131279_cache-entities-contribute-7.6' into cache-entities-contribute-main

Commit:
add cache expiry time

Commit:
cache RelationshipDAO#findByItem as well

bram-atmire reviewed (changes requested):
Good performance improvement. EntityType and RelationshipType are ideal L2 cache candidates: rarely modified, queried constantly. `NONSTRICT_READ_WRITE` is the right strategy, and the 1 hour TTL with reasonable heap sizes (100/500 entries) is sensible.

The `UPPER(label)` index on entity_type is a nice touch, matching the case-insensitive query pattern.

One thing to fix: the migration version is `V7.6_2025.05.20` but this PR targets `main` (DSpace 10). Should be `V10.0_*` to match the convention on this branch. `V7.6` would be appropriate for the backport to `dspace-7_x`.

CI is green but the run is old. Would be good to rebase and get a fresh run against current main.

Commit:
Merge branch 'main' into cache-entities-contribute-main

Commit:
update date on migration files

bram-atmire reviewed (changes requested):
See comment below

bram-atmire reviewed (commented):
*Review context: main branch, Hibernate 6.4.10, Spring Boot 3.5.x*

## Summary

Good approach to cache `EntityType` and `RelationshipType` -- these are effectively configuration data with high read-to-write ratios, and the implementation follows existing DSpace caching patterns (EhCache + `@Cacheable` + query hints). The Flyway migrations and cache sizing are reasonable.

However, there is one change that might **degrade** rather than improve performance, and an index inconsistency between H2 and PostgreSQL.

## Issues

### `Relationship.findByItem` query cache might hurt performance

In `RelationshipDAOImpl.java`, the `findByItem` query is changed from `cacheable=false` to `cacheable=true`. Unlike `EntityType` and `RelationshipType`, the `Relationship` entity is **not** `@Cacheable` (correctly so -- relationships are mutable, per-item data).

Enabling the query cache for a non-L2-cached entity likely causes two problems:

1. **N+1 SELECTs on cache hit**: The query cache stores entity IDs, not full objects. On a cache hit, Hibernate would need to load each `Relationship` individually since it's not in the L2 cache. For an item with 20 relationships, a "cached" query could trigger 20 individual SELECTs instead of 1 query.

2. **Frequent invalidation**: Any INSERT/UPDATE/DELETE to the `relationship` table invalidates ALL cached queries for that table. Since relationships likely change frequently (item creation, versioning, bulk imports), the cache hit rate would probably be low.

This might be worth reverting to `false`. The EntityType and RelationshipType query caching is correct and beneficial.

### H2 vs PostgreSQL index mismatch

The H2 migration creates:
```sql
CREATE INDEX entity_type_label_upper_idx ON entity_type (label);
```

While the PostgreSQL migration creates:
```sql
CREATE INDEX entity_type_label_upper_idx ON entity_type (UPPER(label));
```

If the DAO query uses `UPPER(label)` for case-insensitive lookup, the H2 index likely won't be used (it's a plain B-tree on `label`, not on `UPPER(label)`). Integration tests (which use H2) wouldn't exercise the same index behavior as production. It might be worth making both consistent.

### Minor: missing cache comments

Existing cache entries in `hibernate-ehcache-config.xml` include XML comments explaining why the entity is cached. The new `RelationshipType` and `EntityType` blocks should ideally follow the same pattern.

### Enhancement: consider `READ_ONLY` for EntityType

Entity types are effectively immutable configuration data. `CacheConcurrencyStrategy.READ_ONLY` would likely be more efficient (no lock overhead, no invalidation cost) than `NONSTRICT_READ_WRITE`. Worth evaluating whether EntityTypes are ever updated in production.

## Highlights

- Clean, minimal change following existing patterns
- Correct identification of EntityType and RelationshipType as cache candidates
- Good Flyway migration practice with separate H2/PostgreSQL scripts
- Reasonable cache sizing (100 EntityType, 500 RelationshipType) and 1-hour TTL
- No linked issue ticket -- consider creating one for traceability
- No automated tests for cache behavior -- the manual benchmarking script is documented but an integration test verifying cache hits would strengthen confidence

Commit:
undo cache findByItem query in RelationshipDAOImpl

Commit:
add comments to cache configs

AbhinavS96 commented:
@bram-atmire 

I checked the N+1 issue and indeed the base query is not expensive so caching does not make sense. I have reverted that change. 

Regarding the H2 migration and CacheConcurrencyStrategy, both of those were done in such a way to resolve issues in the CI. I have added comments to hibernate-ecache-config.xml.

Commit:
add comments to h2 migration

AbhinavS96 commented:
created #12511 retroactively

bram-atmire reviewed (approved):
Thanks for the follow-ups, @AbhinavS96 — everything from the previous review is now addressed. 🎉

- **✅ `Relationship.findByItem` query cache** — Reverted (`9c57f39`). Agreed it isn't worth caching given the base query is cheap and `Relationship` isn't `@Cacheable` (would have risked N+1 loads on cache hits). This was the important one.
- **✅ H2 index comment** — The new comment in the H2 migration explaining that H2 doesn't support `UPPER()` and that it has no meaningful impact in a test environment is exactly what was needed for future readers.
- **✅ Cache comments** — `RelationshipType`/`EntityType` blocks in `hibernate-ehcache-config.xml` now follow the existing convention.
- **✅ Issue ticket** — Thanks for creating #12511 for traceability.
- **✅ Migration naming & rebase** — `V7.6_2026.03.25__...` is correct for backporting per @tdonohue's note, and CI is green on a fresh run against current main.

`NONSTRICT_READ_WRITE` over my earlier `READ_ONLY` suggestion is fine — that was only an enhancement, and you've noted the CI rationale.

One tiny optional nit: consider adding `Fixes #12511` to the PR description so the issue auto-closes on merge (right now #12511 isn't formally linked as a closing reference). Not a blocker.

LGTM — approving. 👍

tdonohue reviewed (approved):
👍 Thanks @AbhinavS96 ! The code looks good to me.  I also tested this locally and found no change in behavior (I don't have a setup to verify the performance improvements, but I trust this will provide better performance...all I wanted to very was that it didn't change anything negatively)

tdonohue commented:
@AbhinavS96 : It looks like this PR was only able to be auto-backported to `dspace-7_x`.  Could you please create a separate backport PR for 8.x and 9.x against the `dspace-8_x` and `dspace-9_x` branches, respectively?  That way we get this fix into all supported releases.
